pkg/scheduler/plugins/multihierarchyelasticquota/quota_tree_wrapper.go

```diff
@@ -0,0 +1,476 @@
+	"sync"
+
+	"github.com/koordinator-sh/koordinator/pkg/util"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	v12 "k8s.io/apiserver/pkg/quota/v1"
+	"k8s.io/klog/v2"
+)
+
+type quotaNode struct {
+	quotaName         QuotaNameType
+	request           int64
+	sharedRatio       int64
+	minQuota          int64 //ensure的概念
+	runtimeQuota      int64
+	allowLentResource bool
+}
+
+func (node *quotaNode) String() string {
+	return fmt.Sprintf("qName:%s, req:%v, sharedRatio:%v, minQuota:%v, runtimeQuota:%v",
+		node.quotaName, node.request, node.sharedRatio, node.minQuota, node.runtimeQuota)
+}
+
+func NewQuotaNode(quotaName QuotaNameType, sharedRatio, request, min int64, allowLentResource bool) *quotaNode {
+	if sharedRatio == 0 {
+		return nil
+	}
+
+	return &quotaNode{
+		quotaName:         quotaName,
+		request:           request,
+		sharedRatio:       sharedRatio,
+		minQuota:          min,
+		runtimeQuota:      0,
+		allowLentResource: allowLentResource,
+	}
+}
+
+type quotaMapType map[QuotaNameType]v1.ResourceList
+
+//QuotaTreeWrapper 多资源维度quota runtime管理
+type QuotaTreeWrapper struct {
+	// When heightVersion is updated, this version++
+	heightVersion int64
+	// resource key vector
+	resourceKeys map[string]struct{}
+	// request limit for each group
+	groupReqLimit quotaMapType
+	// 去掉了quotaTree的包装，直接暴露quotaInfo，用作计算runtimeQuota
+	quotaInfoMap map[string]map[QuotaNameType]*quotaNode
+	// init equal mTotalResourceNoSys,then if req[id] > minQuota[id], Resource' -= minQuota[id]
+	totalResource v1.ResourceList //表示该层级的所有quota组可瓜分的资源，含(min)
+	lock          sync.RWMutex
+
+	treeName  string
+	quotaType string
+}
+
+func NewQuotaTreeWrapper(treeName string, quotaType string) *QuotaTreeWrapper {
+	return &QuotaTreeWrapper{
+		heightVersion: 1,
+		resourceKeys:  make(map[string]struct{}),
+		groupReqLimit: make(quotaMapType),
+		quotaInfoMap:  make(map[string]map[QuotaNameType]*quotaNode),
+		totalResource: util.NewResourceList(),
+		treeName:      treeName,
+		quotaType:     quotaType,
+	}
+}
+
+func (qtw *QuotaTreeWrapper) UpdateResourceKeys(resourceKeys map[string]struct{}) {
+	newResourceKey := make(map[string]struct{})
+	for resKey := range resourceKeys {
+		newResourceKey[resKey] = struct{}{}
+	}
+
+	qtw.lock.Lock()
+	defer qtw.lock.Unlock()
+
+	qtw.resourceKeys = newResourceKey
+	qtw.updateQuotaTreeDimensionByResourceKeys()
+}
+
+func (qtw *QuotaTreeWrapper) updateQuotaTreeDimensionByResourceKeys() {
+	//lock outside
+	for resKey := range qtw.quotaInfoMap {
+		if _, exist := qtw.resourceKeys[resKey]; !exist {
+			delete(qtw.quotaInfoMap, resKey)
+		}
+	}
+
+	for resKey := range qtw.resourceKeys {
+		if _, exist := qtw.quotaInfoMap[resKey]; !exist {
+			qtw.quotaInfoMap[resKey] = make(map[QuotaNameType]*quotaNode)
+		}
+	}
+}
+
+func (qtw *QuotaTreeWrapper) DeleteOneGroup(quotaInfo *QuotaInfo) {
+	// TODO 需要注意参数传递过来的是指针，在内部访问的时候需要保证线程安全
+	qtw.lock.Lock()
+	defer qtw.lock.Unlock()
+
+	klog.V(5).Infof("DeleteOneGroup before"+
```

Comment:
the log is too long... Recommend the following:
```go
if klog.V(5).Enabled {
   //....
}
```

Or abstract logging as a function... I noticed there is a lot of logging code like this. Please keep the code clean and let's focus on the most important logic.

Reply:
done